### PR TITLE
Databricks client implementation

### PR DESCRIPTION
**Pull request by serena-ruan**
Initial version of UC function client
- BaseFunctionClient as the basic client
- DatabricksFunctionClient for connecting with databricks managed UC

Main functionalities:
- create_function
- get_function
- list_functions
- execute_function

Tests added in test_client function can be executed; test_databricks_client is tested locally due to the limitation of authentication and warehouse usage; 

NOTE: create_function with function_info is pending on a bug on the API https://docs.databricks.com/api/workspace/functions/create, so please ignore this part for now.

### CODE REVIEW

File: tests/functions/test_databricks_client.py
Comment (serena-ruan) [inline tests/functions/test_databricks_client.py]: Pls ignore this test for now as it's blocked by an API bug

File: pyproject.toml
Comment (harupy) [inline pyproject.toml]:
```suggestion
requires-python = ">=3.9"
```

3.8 will reach EOL soon.

File: pyproject.toml
Comment (harupy) [inline pyproject.toml]:
```suggestion
name = "unitycatalog-functions"
```

should this be plural?

Comment (harupy) [reply pyproject.toml]: btw is this a name space package?

Comment (serena-ruan) [reply pyproject.toml]: Yea this name is pending to be changed, I'm waiting for Xiangrui's response :)

File: README.md
Comment (harupy) [inline README.md]: what is a production repo?

File: pyproject.toml
Comment (serena-ruan) [reply pyproject.toml]: lol just found it's Oct 2024!

File: README.md
Comment (serena-ruan) [reply README.md]: Maybe just 'repo' 😆

File: unitycatalog/functions/client.py
Comment (harupy) [inline unitycatalog/functions/client.py:1]: This is not a blocker, but I'd recommend using the `src` layout. https://packaging.python.org/en/latest/discussions/src-layout-vs-flat-layout/ illustrates why the src layout makes sense.

File: pyproject.toml
Comment (harupy) [inline pyproject.toml]:
```suggestion
description = "Official python library for Unity Catalog functions support"
```

File: pyproject.toml
Comment (harupy) [inline pyproject.toml]:
```suggestion
  "ruff==x.y",
```

let's pin ruff.

File: README.md
Comment (harupy) [reply README.md]: I'd say "Install from the source"

File: unitycatalog/functions/client.py
Comment (harupy) [inline unitycatalog/functions/client.py]:
```suggestion
    def get_function(self, function_name: str, **kwargs: Any) -> Any:
```

for easier and faster typing.

File: README.md
Comment (harupy) [inline README.md:25]: Is `databricks-sdk` lacking this functionality?

File: unitycatalog/functions/client.py
Comment (harupy) [inline unitycatalog/functions/client.py]: do we need to use pop? can we use get?

File: unitycatalog/functions/client.py
Comment (harupy) [inline unitycatalog/functions/client.py]: how is this function used?

File: unitycatalog/functions/databricks.py
Comment (harupy) [inline unitycatalog/functions/databricks.py]:
```suggestion
def column_type_to_python_type(column_type: ColumnTypeName) -> Any:
```

File: unitycatalog/functions/client.py
Comment (harupy) [inline unitycatalog/functions/client.py]:
```suggestion
```

can this be removed?

File: unitycatalog/functions/databricks.py
Comment (harupy) [inline unitycatalog/functions/databricks.py]: Can we add a `list_functions` instead?

File: unitycatalog/functions/databricks.py
Comment (harupy) [inline unitycatalog/functions/databricks.py]: do we need to override this?

File: unitycatalog/functions/databricks.py
Comment (harupy) [inline unitycatalog/functions/databricks.py]:
```suggestion
                for _ in range(max_retries):
                    time.sleep(2**retry_cnt)
                    _logger.info(f"Retry times: {retry_cnt}")
                    response = self.client.statement_execution.get_statement(statement_id)
                    if response.status is None or response.status.state != StatementState.PENDING:
                        break
```

Can we use a loop instead?

File: unitycatalog/functions/databricks.py
Comment (harupy) [inline unitycatalog/functions/databricks.py]: can we include `sql_body` in the message?

File: unitycatalog/functions/client.py
Comment (BenWilson2) [reply unitycatalog/functions/client.py:1]: +1 to this

File: pyproject.toml
Comment (BenWilson2) [reply pyproject.toml]: Let's capitalize `Python` too :) https://devguide.python.org/documentation/style-guide/#specific-words

File: README.md
Comment (BenWilson2) [inline README.md:2]: Let's change references to be `Unity Catalog` (similar to Databricks Docs)

File: README.md
Comment (BenWilson2) [inline README.md]:
```suggestion
The library includes clients for interacting with both the Open-Source Unity Catalog server and the Databricks-managed Unity Catalog service.
```

File: README.md
Comment (BenWilson2) [inline README.md]:
```suggestion
To use Databricks-managed Unity Catalog with this package, follow the [instructions](https://docs.databricks.com/en/dev-tools/cli/authentication.html#authentication-for-the-databricks-cli) to authenticate to your Workspace and ensure that your access token has workspace-level privilege for managing UC functions.
```

File: unitycatalog/functions/databricks.py
Comment (BenWilson2) [inline unitycatalog/functions/databricks.py]: Is there a similar mechanism for the OSS UC catalog to retrieve the type conversion fields? If so, should we externalize an abstract implementation of this into a common util module?

File: unitycatalog/functions/databricks.py
Comment (BenWilson2) [inline unitycatalog/functions/databricks.py]: Similar comment as above. It might be worthwhile to abstract this as a util that both Databricks and UC can use with conditional imports for the `ColumnTypeName` entries that are required for type validation. 
I like the idea of this module having only the client interface APIs in it.

File: README.md
Comment (serena-ruan) [reply README.md:25]: Ah actually, databricks-sdk supports executing command against a cluster, it doesn't support serverless. Let me add support of using cluster_id with databricks-sdk in a follow-up PR and then update this section :)

File: unitycatalog/functions/client.py
Comment (serena-ruan) [reply unitycatalog/functions/client.py]: Ah it's because I check if there're leftovers in the parameters on line 72, and raise ValueError if there are.

File: unitycatalog/functions/client.py
Comment (serena-ruan) [reply unitycatalog/functions/client.py]: Yep!

File: unitycatalog/functions/client.py
Comment (serena-ruan) [reply unitycatalog/functions/client.py]: It's not used yet, it will be used in langchainToolkit, something like below when defining the tool's func:
```
def func(*args: Any, **kwargs: Any) -> str:
        client = get_uc_function_client()
        args_json = json.loads(json.dumps(kwargs, default=str))
        result = client.execute_function(
            function_name=func_name,
            parameters=args_json,
        )
        return result.to_json()
```

File: unitycatalog/functions/databricks.py
Comment (serena-ruan) [reply unitycatalog/functions/databricks.py]: Make sense! OSS UC uses Literal["Boolean", ...] with the same value as this mapping here, this is Enum so we can also get the string value!

File: unitycatalog/functions/databricks.py
Comment (serena-ruan) [reply unitycatalog/functions/databricks.py]: I'm overriding it just for the docstring, is there any other ways to do it?

File: src/unitycatalog/functions/client.py
Comment (harupy) [inline src/unitycatalog/functions/client.py]: can we add docstring for the arguments?

File: src/unitycatalog/functions/client.py
Comment (harupy) [inline src/unitycatalog/functions/client.py:46]: Any constraint on the return type? Can this function return any object?

File: src/unitycatalog/functions/databricks.py
Comment (harupy) [inline src/unitycatalog/functions/databricks.py]:
```suggestion
        self.client = client or get_default_databricks_workspace_client()
```

File: tests/functions/test_databricks_client.py
Comment (harupy) [inline tests/functions/test_databricks_client.py:39]: Even if it's in an early development phase, I'd recommend adding a minimum CI to prevent silly mistakes. We don't need to test everything. Just run the following

1. Install the package
2. Run pytest

File: unitycatalog/functions/databricks.py
Comment (harupy) [reply unitycatalog/functions/databricks.py]: got it, didn't notice that.

File: unitycatalog/functions/client.py
Comment (harupy) [reply unitycatalog/functions/client.py]: I see, I think I'd use the following:

```python
if keys := parameters.keys() - {p.name for p in input_params.parameters}:
    ...
```

File: pyproject.toml
Comment (harupy) [inline pyproject.toml]:
```suggestion
target-version = "py39"
```

File: src/unitycatalog/functions/client.py
Comment (serena-ruan) [reply src/unitycatalog/functions/client.py:46]: The return types for UC OSS and Databricks UC are different, and I don't want to depend on these two different packages in this file, so I just set Any

File: tests/functions/test_databricks_client.py
Comment (serena-ruan) [reply tests/functions/test_databricks_client.py:39]: Emm it won't work in the CI now, because it needs authentication to a databricks workspace and a warehouse & cluster for testing everything, but I always run the tests locally to make sure they pass before pushing.
image

Comment (harupy) [reply tests/functions/test_databricks_client.py:39]: Got it. I think you can still test the package installation.

Comment (harupy) [reply tests/functions/test_databricks_client.py:39]: > it needs authentication to a databricks workspace and a warehouse

Can we use mocks? or you could create a fake Databricks using FastAPI.

Comment (harupy) [reply tests/functions/test_databricks_client.py:39]: We don't need to solve this right now, but imagine an external contributor files a PR in the future. He/she can't access secrets = he/she can't run tests for Databricks.

Relying on external resources like Databricks has a risk. If it's down, we can't run tests, which slows us down.

Comment (serena-ruan) [reply tests/functions/test_databricks_client.py:39]: I'll mark those tests involving databricks integration to depend on an environment variable and skip in the normal tests. The rest can be run directly -- mainly for checking the logics.

File: .github/workflows/main.yml
Comment (harupy) [inline .github/workflows/main.yml]: nit: indent

File: .github/workflows/main.yml
Comment (harupy) [inline .github/workflows/main.yml]: Is `pip install .` missing?

File: src/unitycatalog/functions/client.py
Comment (harupy) [inline src/unitycatalog/functions/client.py]:
```suggestion
    ) -> PagedList[Any]:
```

**Review by harupy — APPROVED** [review]
LGTM

File: src/unitycatalog/functions/utils.py
Comment (harupy) [inline src/unitycatalog/functions/utils.py]:
```suggestion
def validate_full_function_name(function_name: str) -> Tuple[str, str, str]:
```

I'd return a tuple or a named tuple. `List[str]` basically means "not sure how many elements it has."

File: .github/workflows/test-package-build.yml
Comment (harupy) [inline .github/workflows/test-package-build.yml]: We can remove these steps. MLflow includes lots of non-python files so inspecting them makes sense, but this package doesn't.

**Comment by serena-ruan** [comment]
Thanks so much for the detailed review @harupy !!! I'll merge this first, and if there're any potential issues I can fix later

**Pull request merged**
